# Sprint 2 – Notebook 1: Introduction to Statistics

In [1]:
import numpy as np
import pandas as pd
from scipy import stats

np.random.seed(42)

# Small structured dataset
df = pd.DataFrame({
    "CustomerID": [1, 2, 3, 4, 5, 6, 7, 8],
    "Age": [25, 32, 47, 51, 23, 36, 29, 41],
    "City": ["Chennai", "Madurai", "Thanjavur", "Chennai", "Madurai", "Trichy", "Chennai", "Trichy"],
    "Salary": [30000, 52000, 61000.5, 75000, 28000, 58000, 41000, 66000],
    "NumOrders": [2, 5, 7, 9, 1, 6, 3, 8],
    "Purchased": ["No", "Yes", "Yes", "Yes", "No", "Yes", "No", "Yes"],
})

# Population of 100,000 customers' monthly spend
population = np.random.normal(loc=5000, scale=1200, size=100_000)

## 1. What is Statistics?

Statistics is the science of collecting, organizing, summarizing, analyzing and interpreting data so that we can make decisions in the presence of uncertainty.

In simple words: data is raw, and statistics turns it into meaning.

**Real-world analogy:** A doctor checks a few drops of your blood, not all of it, and still concludes about your whole body. Statistics gives the rules for making such conclusions safely.

## 2. Why Statistics is Important in AI/ML

- Helps us understand data before modeling (average, spread, patterns)
- Helps detect problems: missing values, outliers, bias
- Helps choose features and check relationships
- Helps evaluate models (accuracy, confidence, significance)
- Many ML algorithms are statistical at their core (Linear Regression, Naive Bayes, PCA)

## 3. Statistics vs Machine Learning

| Aspect | Statistics | Machine Learning |
|---|---|---|
| Main goal | Understand and explain data, inference | Predict accurately on new data |
| Approach | Model assumptions, hypothesis tests | Algorithms learned from data |
| Focus | Why does this happen? | What will happen next? |
| Data size | Often small/medium samples | Often large datasets |
| Example | Does the ad campaign increase sales? (p-value) | Predict which customer will buy |

They overlap heavily: ML uses statistical ideas, and statistics uses computation.

## 4. Descriptive Statistics

Descriptive statistics summarizes the data we already have (mean, median, charts).

**Example:** The average marks of my class of 40 students is 72.
**AI/ML use:** EDA (Exploratory Data Analysis) and data profiling before modeling.

In [2]:
# Descriptive: summarize the data in hand
print(df[["Age", "Salary", "NumOrders"]].describe().round(2))

         Age    Salary  NumOrders
count   8.00      8.00       8.00
mean   35.50  51375.06       5.12
std    10.17  16986.91       2.90
min    23.00  28000.00       1.00
25%    28.00  38250.00       2.75
50%    34.00  55000.00       5.50
75%    42.50  62250.38       7.25
max    51.00  75000.00       9.00


**Explanation:** `describe()` gives count, mean, std, min, quartiles and max for each numeric column.
**Interpretation:** Average age is about 35.5, and salary ranges from 28,000 to 75,000. We are only describing these 8 customers, not making claims about anyone else.

## 5. Inferential Statistics

Inferential statistics uses a sample to draw conclusions about a larger population.

**Example:** From 500 surveyed voters, we estimate how the whole state will vote.
**AI/ML use:** A/B testing, model comparison, deciding whether an improvement is real or just luck.

In [3]:
# Inferential: estimate population mean from a sample using 95% confidence interval
sample_spend = np.random.choice(population, size=200, replace=False)
mean = sample_spend.mean()
sem = stats.sem(sample_spend)
ci = stats.t.interval(0.95, df=len(sample_spend) - 1, loc=mean, scale=sem)

print("Sample mean               :", round(mean, 2))
print("95% CI for population mean:", tuple(round(x, 2) for x in ci))
print("True population mean      :", round(population.mean(), 2))

Sample mean               : 5028.66
95% CI for population mean: (np.float64(4862.05), np.float64(5195.28))
True population mean      : 5001.16


**Explanation:** We use only 200 customers to estimate a range where the population mean likely lies.
**Interpretation:** If we repeated this sampling many times, about 95% of such intervals would contain the true mean. Here the true mean falls inside the interval.

| | Descriptive | Inferential |
|---|---|---|
| Purpose | Describe | Generalize / predict |
| Data used | Whole dataset in hand | Sample -> population |
| Tools | Mean, median, graphs | Hypothesis tests, confidence intervals |

## 6. Population
The complete group we want to study (all customers of a bank).

## 7. Sample
A smaller subset of the population that we actually observe (1,000 customers).

## 8. Census
Collecting data from every member of the population (India's national census).

## 9. Sampling
The process of selecting a sample from the population.

**Why sample?** Census is costly, slow and sometimes impossible (testing every bulb until it fails would destroy them all).

**AI/ML link:** Training data is a *sample*. The real world is the *population*. A model is good only if the sample represents the population.

In [6]:
# Census: use the entire population
print("Population mean (census, true value):", round(population.mean(), 2))

# Sampling: pick 200 random customers
sample = np.random.choice(population, size=200, replace=False)
print("Sample mean (estimate)              :", round(sample.mean(), 2))
print("Difference                          :", round(abs(population.mean() - sample.mean()), 2))

# Different samples give slightly different results
for i in range(3):
    s = np.random.choice(population, size=200, replace=False)
    print(f"Sample {i+1} mean:", round(s.mean(), 2))

Population mean (census, true value): 5001.16
Sample mean (estimate)              : 5051.75
Difference                          : 50.59
Sample 1 mean: 5069.01
Sample 2 mean: 4999.76
Sample 3 mean: 4948.31


**Explanation:** The full population calculation acts as a census. Then we draw random samples of 200 and compare.
**Interpretation:** Sample means are close to the population mean but not identical, and they change from sample to sample (sampling variation). A bigger, well-chosen sample gives a closer estimate.
**AI/ML use:** Your training set is a sample, so if it is small or biased, the model learns the real world wrongly. That is why we do train/test split and validation.

## 10. Variable
Any characteristic that can take different values (age, salary, city).

## 11. Feature
A variable used as an input to an ML model. The variable we predict is the **target/label**.

## 12. Observation
One row, one record (one customer).

## 13. Dataset
The full collection of observations and variables (the whole table).

| CustomerID | Age | City | Salary | Purchased |
|---|---|---|---|---|
| 1 | 25 | Chennai | 30000 | No |
| 2 | 32 | Madurai | 52000 | Yes |

Each row = observation, Age/City/Salary = features, Purchased = target, whole table = dataset.

In [7]:
print("Dataset shape (observations, variables):", df.shape)
print("Number of observations (rows)  :", df.shape[0])
print("Number of variables (columns)  :", df.shape[1])

# Separate features and target
features = df[["Age", "City", "Salary", "NumOrders"]]
target = df["Purchased"]
print("\nFeatures:", list(features.columns))
print("Target  :", target.name)

# One observation
print("\nOne observation (row 0):")
print(df.iloc[0])
df

Dataset shape (observations, variables): (8, 6)
Number of observations (rows)  : 8
Number of variables (columns)  : 6

Features: ['Age', 'City', 'Salary', 'NumOrders']
Target  : Purchased

One observation (row 0):
CustomerID          1
Age                25
City          Chennai
Salary        30000.0
NumOrders           2
Purchased          No
Name: 0, dtype: object


,CustomerID,Age,City,Salary,NumOrders,Purchased
0,1,25,Chennai,30000.0,2,No
1,2,32,Madurai,52000.0,5,Yes
2,3,47,Thanjavur,61000.5,7,Yes
3,4,51,Chennai,75000.0,9,Yes
4,5,23,Madurai,28000.0,1,No
5,6,36,Trichy,58000.0,6,Yes
6,7,29,Chennai,41000.0,3,No
7,8,41,Trichy,66000.0,8,Yes


**Explanation:** `df.shape` gives rows and columns. We split the columns into features (input) and target (output), and print one row as a single observation.
**Interpretation:** 8 observations and 6 variables. `Purchased` is the target, and the others are features.

## 14. Quantitative Data
Numerical and measurable (age, salary).

## 15. Qualitative Data
Categories or labels (city, gender, colour). Can be nominal (no order, like city) or ordinal (has order, like Low/Medium/High).

## 16. Continuous Data
Any value in a range, decimals possible (height 172.5 cm, temperature).

## 17. Discrete Data
Countable whole numbers (number of children, number of orders).

**AI/ML use:** Numeric features need scaling. Qualitative features need encoding (one-hot, label encoding) before training.

In [8]:
print(df.dtypes)

quantitative = df.select_dtypes(include=[np.number]).columns.tolist()
qualitative = df.select_dtypes(exclude=[np.number]).columns.tolist()
print("\nQuantitative:", quantitative)
print("Qualitative :", qualitative)

# Discrete vs Continuous (all whole numbers -> discrete)
print()
for col in quantitative:
    kind = "Discrete" if (df[col] % 1 == 0).all() else "Continuous"
    print(f"{col:12s} -> {kind}")

CustomerID      int64
Age             int64
City              str
Salary        float64
NumOrders       int64
Purchased         str
dtype: object

Quantitative: ['CustomerID', 'Age', 'Salary', 'NumOrders']
Qualitative : ['City', 'Purchased']

CustomerID   -> Discrete
Age          -> Discrete
Salary       -> Continuous
NumOrders    -> Discrete


**Explanation:** Numeric columns are separated from text columns, then we check whether numeric values are all whole numbers.
**Interpretation:** `Salary` has a decimal (61000.5), so it is continuous. `Age` and `NumOrders` are whole-number counts, so discrete. `City` and `Purchased` are qualitative.

## 18. Structured Data
Organized in rows and columns (SQL table, CSV). Goes to classic ML (regression, XGBoost).

## 19. Unstructured Data
No fixed format (images, audio, emails, tweets). Needs deep learning (CNN for images, transformers for text).

In [9]:
# Structured: fixed rows and columns
print("STRUCTURED (DataFrame):")
print(df.head(3))

# Unstructured: free text, no fixed schema
reviews = [
    "Delivery was very fast, loved the product!",
    "Packaging was damaged and support did not reply.",
]
print("\nUNSTRUCTURED (raw text):")
for r in reviews:
    print("-", r, "| words:", len(r.split()))

STRUCTURED (DataFrame):
   CustomerID  Age       City   Salary  NumOrders Purchased
0           1   25    Chennai  30000.0          2        No
1           2   32    Madurai  52000.0          5       Yes
2           3   47  Thanjavur  61000.5          7       Yes

UNSTRUCTURED (raw text):
- Delivery was very fast, loved the product! | words: 7
- Packaging was damaged and support did not reply. | words: 8


**Explanation:** The DataFrame is structured because every row has the same columns. The reviews are unstructured text, and we can only extract simple things like word count until we process them (NLP).
**Interpretation:** Unstructured data must be converted into numbers (tokens, embeddings) before ML can use it.